test.c

```diff
@@ -59,9 +67,28 @@ void do_accept(struct utcp_connection *nc, uint16_t port) {
 ssize_t do_send(struct utcp *utcp, const void *data, size_t len) {
 	int s = *(int *)utcp->priv;
 	outpktno++;
-	if(outpktno < dropto && outpktno >= dropfrom && drand48() < dropout)
-		return len;
+	if(outpktno >= dropfrom && outpktno < dropto) {
+		if(drand48() < dropout)
+			return len;
+		if(!reorder_data && drand48() < reorder) {
+			reorder_data = malloc(len);
```

Comment:
On Thu, Oct 29, 2015 at 03:48:00AM -0700, vollkorn1982 wrote:

> > @@ -59,9 +67,28 @@ void do_accept(struct utcp_connection *nc, uint16_t port) {
> >  ssize_t do_send(struct utcp *utcp, const void *data, size_t len) {
> >  	int s = *(int *)utcp->priv;
> >  	outpktno++;
> > -	if(outpktno < dropto && outpktno >= dropfrom && drand48() < dropout)
> > -		return len;
> > +	if(outpktno >= dropfrom && outpktno < dropto) {
> > +		if(drand48() < dropout)
> > +			return len;
> > +		if(!reorder_data && drand48() < reorder) {
> > +			reorder_data = malloc(len);
> 
> Can you be sure `len` will never be 0? Because with `len = 0` the returned pointer is implementation defined.

Yes, UTCP will never send zero length packets. But the return value from
malloc() should be checked here in any case, even if len > 0.

-- 
Met vriendelijke groet / with kind regards,
     Guus Sliepen <guus@tinc-vpn.org>